### PR TITLE
issue 11391: Support for multiple backends

### PR DESCRIPTION
**Pull request by VRanga000**
Adds the ability to configure multiple backends (output plugins) for fluent-bit

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes # 11391

[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by kfox1111** [comment]
There's always a tradeoff between hitting the common cases to make them easy, and features that require more and more code to map values to config making the code base hard to maintain.

All config options can already be set by using your own configmap or by overriding rawConfing.

This feels like it crosses the line and makes it harder to use in the common case, and adds even more config code.

Is configmap uploading or overriding rawConfig sufficient for your need?

**Comment by VRanga000** [comment]
> There's always a tradeoff between hitting the common cases to make them easy, and features that require more and more code to map values to config making the code base hard to maintain.
> 
> All config options can already be set by using your own configmap or by overriding rawConfing.
> 
> This feels like it crosses the line and makes it harder to use in the common case, and adds even more config code.
> 
> Is configmap uploading or overriding rawConfig sufficient for your need?
Thanks for the response. I am relatively new to fluentbit, but it seems like the mechanism to use multiple OUTPUTs would be commonly used. The current chart code does not allow for that (easily) without writing the entire config (as you suggested) by overriding rawConfig. I don't think my changes are adding more code - but they are adding more flexibility by allowing multiple backends to be selected - making common cases more convenient to cover. I don't think this adds more code per section going forward either.. could you take another look at the diff and see? Thanks!

**Comment by Towmeykaw** [comment]
There needs to be a strong benefit for this as this would be a breaking change. Adding output as rawConfig would be done something like this 
```
@INCLUDE fluent-bit-output.conf

[OUTPUT]
    Name  stdout
    Match *
```

**Comment by VRanga000** [comment]
> There needs to be a strong benefit for this as this would be a breaking change. Adding output as rawConfig would be done something like this
> 
> ```
> @INCLUDE fluent-bit-output.conf
> 
> [OUTPUT]
>     Name  stdout
>     Match *
> ```

Agreed that just the ability to add stdout is not compelling enough. My issue started out that way. 
But the main benefit of this PR is to allow *multiple* backends to be enabled.
For example, you can now enable both stdout and es backends via the values.yaml file:
```
backend:
  file:
    enabled: false
  stdout:
    enabled: true
  forward:
    enabled: false
  es:
    enabled: true
```
which results in the following config:
```
fluent-bit-output.conf:
----
[OUTPUT]
    Name stdout
    Match *
[OUTPUT]
    Name  es
    Match *
    Host  myesdomain.us-east-1.es.amazonaws.com
    Port  443
    Logstash_Format On
```
whereas before this PR, you could only enable 1 backend (e.g.):
```
backend:
  type: es
```
I can create a new issue (and perhaps close the current one) so that it's less confusing.

**Comment by VRanga000** [comment]
Created new issue ( #11391 ) for this PR - to clear up confusion. The stdout issue (#11331 ) can be closed/abandoned, if needed.

**Comment by kfox1111** [comment]
my objection was more towards having multiple output sections enabled at once.

It makes it harder to switch backends as you have to enable the one you want and disable the other one, rather then just picking one.

The way I've always envisioned this chart to be used would be fluent-bit being used purely as a log shipper. pull off the logs off the host and ship it to something like fluentd to slice/dice/enhance/send to other backends, etc. As fluent-bit runs on every host, it should be as light weight as possible to not effect workload of the system and you can deal with all the processing in a deployment where it is easier to track/scale/upgrade, etc.

**Comment by VRanga000** [comment]
> my objection was more towards having multiple output sections enabled at once.
> 
> It makes it harder to switch backends as you have to enable the one you want and disable the other one, rather then just picking one.
> 
> The way I've always envisioned this chart to be used would be fluent-bit being used purely as a log shipper. pull off the logs off the host and ship it to something like fluentd to slice/dice/enhance/send to other backends, etc. As fluent-bit runs on every host, it should be as light weight as possible to not effect workload of the system and you can deal with all the processing in a deployment where it is easier to track/scale/upgrade, etc.

Thanks for the response. I don't see how it makes it harder. e.g. Let's say you've configured both stdout and elastic search as backends (during initial test phase). Now you want to disable the stdout backend and keep the elasticsearch one. All you have to do is this:

```
helm upgrade <fluent-bit-release> --set backend.stdout.enabled=false 
```
And if you could just as easily do this in a ci pipeline by updating your custom values.yaml file (for example). As far as your other objection of fluent-bit being lightweight, I see your point - that you probably don;t want to enable multiple backends - but there are still cases (like the one I mentioned) - where you might want to (at least temporarily) enable multiple backends. In any case, my changes don't make it necessary for you to enable multiple backends. You can still just enable exactly one. So it does not make existing functionality any more tedious/difficult, but adds more flexibility.

File: stable/fluent-bit/README.md [hunk 1]
```diff
@@ -29,16 +29,20 @@ The following table lists the configurable parameters of the Fluent-Bit chart an
 
 | Parameter                  | Description                        | Default                 |
 | -----------------------    | ---------------------------------- | ----------------------- |
-| **Backend Selection**      |
-| `backend.type`             | Set the backend to which Fluent-Bit should flush the information it gathers | `forward` |
+| **file Backend**        |
+| `backend.file.enabled`     | Boolean flag to indicate whether to enable the "file" output plugin | `fluentd` |
+| **stdout Backend**        |
+| `backend.stdout.enabled`     | Boolean flag to indicate whether to enable the "stdout" output plugin | `fluentd` |
```
Comment (legal90) [inline stable/fluent-bit/README.md:35]:
```suggestion
| `backend.stdout.enabled`     | Boolean flag to indicate whether to enable the "stdout" output plugin | `true` |
```
It's enabled by default, so the default value should be `true`, I guess. And similarly it should be `false` for other backends here.

File: stable/fluent-bit/values.yaml [hunk 2]
```diff
@@ -24,15 +24,20 @@ metrics:
 trackOffsets: false
 
 backend:
-  type: forward
+  file:
+    enabled: false
+  stdout:
+    enabled: true
   forward:
+    enabled: false
```
Comment (kfox1111) [inline stable/fluent-bit/values.yaml:32]: This goes against the recommendation that fluent-bit should be used as a log shipper. It should default to true.

File: stable/fluent-bit/values.yaml [hunk 3]
```diff
@@ -24,15 +24,20 @@ metrics:
 trackOffsets: false
 
 backend:
-  type: forward
+  file:
+    enabled: false
+  stdout:
+    enabled: true
```
Comment (kfox1111) [inline stable/fluent-bit/values.yaml:30]: enabled: false

File: stable/fluent-bit/templates/config.yaml [hunk 4]
```diff
@@ -68,25 +68,32 @@ data:
...
         Name  file
-        Match *
+        Match {{ .Values.backend.file.match | default "*" }}
         Path /tmp/fluent-bit.log
-{{ else if eq .Values.backend.type "forward" }}
+{{- end }}
+{{- if .Values.backend.stdout.enabled }}
+    [OUTPUT]
+        Name stdout
+        Match {{ .Values.backend.stdout.match | default "*" }}
```
Comment (kfox1111) [inline stable/fluent-bit/templates/config.yaml:80]: Why is stdout is ever a good idea to enable in this context.

It pushes all logs to stdout. it itself is a container. which generates a log file. which fluent-bit reads for logs and displays them.

How do you avoid getting stuck in a loop endlessly spewing logs?

File: stable/fluent-bit/templates/config.yaml [hunk 5]
```diff
@@ -115,10 +122,11 @@ data:
         tls.ca_file /secure/es-tls-ca.crt
 {{- end }}
 {{- end }}
-{{ else if eq .Values.backend.type "splunk" }}
+{{- end }}
+{{- if .Values.backend.splunk.enabled }}
     [OUTPUT]
         Name  splunk
-        Match *
+        Match {{ .Values.backend.splunk.match | default "*" }}
```
Comment (kfox1111) [inline stable/fluent-bit/templates/config.yaml:129]: Making all the matches configurable looks to  be a good change.

**Review by kfox1111 — COMMENTED** [review]
Its still harder to switch as you need to --set backend.es.enabled=true and backend.forward.enabled=false rather then just set backend.type=es

Do you think it is going to be common practice to have a log shipper be deploying to multiple backends at once? I'd rather optimize for the 80% and ensure that it is still possible to do for the rest. rawConfig allows that now.

Do you believe that splitting the stream when one of those backends is having problems will function well? I suspect running two fluent-bit instances with one backend each may be more reliable. then one configured to output to two backends. Having two state files would really help I think.

**Pull request closed without merging**